model/labels/labels_internlabels.go

```diff
@@ -0,0 +1,778 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build internlabels
+
+package labels
+
+import (
+	"bytes"
+	"strings"
+	"sync"
+	"unsafe"
+
+	"github.com/cespare/xxhash/v2"
+	"golang.org/x/exp/slices"
+)
+
+// Labels is implemented by a SymbolTable and string holding name/value
+// pairs encoded as indexes into the table in varint encoding.
+// Names are in alphabetical order.
+type Labels struct {
+	syms *nameTable
+	data string
+}
+
+// Split SymbolTable into the part used by Labels and the part used by Builder.  Only the latter needs the map.
+
+// This part is used by Labels. All fields are immutable after construction.
+type nameTable struct {
+	byNum       []string     // This slice header is never changed, even while we are building the symbol table.
+	symbolTable *SymbolTable // If we need to use it in a Builder.
+}
+
+// SymbolTable is used to map strings into numbers so they can be packed together.
+type SymbolTable struct {
+	mx sync.Mutex
+	*nameTable
+	nextNum int
+	byName  map[string]int
+}
+
+const defaultSymbolTableSize = 1024
+
+func NewSymbolTable() *SymbolTable {
+	t := &SymbolTable{
+		nameTable: &nameTable{byNum: make([]string, defaultSymbolTableSize)},
+		byName:    make(map[string]int, defaultSymbolTableSize),
+	}
+	t.nameTable.symbolTable = t
+	return t
+}
+
+func (t *SymbolTable) Len() int {
+	t.mx.Lock()
+	defer t.mx.Unlock()
+	return len(t.byName)
+}
+
+// ToNum maps a string to an integer, adding the string to the table if it is not already there.
+// Note: copies the string before adding, in case the caller passed part of
+// a buffer that should not be kept alive by this SymbolTable.
+func (t *SymbolTable) ToNum(name string) int {
+	t.mx.Lock()
+	defer t.mx.Unlock()
+	if i, found := t.byName[name]; found {
+		return i
+	}
+	i := t.nextNum
+	if t.nextNum == cap(t.byNum) {
+		// Name table is full; copy to a new one.  Don't touch the existing slice.
+		oldSlice := t.byNum
+		t.nameTable = &nameTable{byNum: make([]string, cap(oldSlice)*2), symbolTable: t}
+		copy(t.nameTable.byNum, oldSlice)
+	}
+	name = strings.Clone(name)
+	t.byNum[i] = name
+	t.byName[name] = i
+	t.nextNum++
+	return i
+}
+
+// ToName maps an integer to a string.
+func (t *nameTable) ToName(num int) string {
```

Comment:
I feel like this should be `ToValue`, but it's opinionated.

Reply:
Given we have explicit names and values for labels, I don't think I should prefer one here.
I can't remember why it's a "name table" and not a "string table".